### PR TITLE
DAOS-16508 csum: retry a few times on checksum mismatch on update

### PR DESCRIPTION
**Pull request by johannlombardi**
Unlike fetch, we return DER_CSUM on update (turned into EIO by dfs) without any retry. We should retry a few times in case it is a transient error.

The patch also prints more information about the actual checksum mismatch.

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/object/cli_obj.c [hunk 1]
```diff
@@ -4760,9 +4760,20 @@ obj_comp_cb(tse_task_t *task, void *data)
 						obj_auxi->tx_uncertain = 1;
 					else
 						obj_auxi->nvme_io_err = 1;
-				} else if (task->dt_result != -DER_NVME_IO) {
-					/* Don't retry update for CSUM & UNCERTAIN errors */
-					obj_auxi->io_retry = 0;
+				} else {
+					if (task->dt_result == -DER_CSUM) {
+						/** Retry a few times on checksum error on update */
+						if (!obj_auxi->csum_retry &&
```
Comment (liuxuezhao) [inline src/object/cli_obj.c:4766]: "!obj_auxi->csum_retry" seems can be removed as it is cleared at above L4689 so it is always true.

File: src/object/obj_internal.h [hunk 2]
```diff
@@ -474,9 +474,10 @@ struct obj_auxi_args {
 					 rebuilding:1,
 					 for_migrate:1;
 	/* request flags. currently only: ORF_RESEND */
-	uint32_t			 flags;
 	uint32_t			 specified_shard;
+	uint16_t			 flags;
```
Comment (liuxuezhao) [inline src/object/obj_internal.h:478]: currently we already used more than 16 bits, 
such as -
5572 >       if (args->extra_flags & DIOF_EC_RECOV_FROM_PARITY)                                          
5573 >       >       obj_auxi->flags |= ORF_EC_RECOV_FROM_PARITY; 

ORF_EC_RECOV_FROM_PARITY = (1 << 19),

File: src/object/obj_internal.h [hunk 3]
```diff
@@ -474,9 +474,10 @@ struct obj_auxi_args {
...
 	uint16_t			 retry_cnt;
+	uint16_t			 csum_retry_cnt;
```
Comment (liuxuezhao) [inline src/object/obj_internal.h:480]: maybe we can directly re-use retry_cnt?

Comment (johannlombardi) [reply src/object/obj_internal.h:480]: the issue is that this counter can be incremented for many reason and we might end up not retrying on checksum error if the RPCs failed many time previously for another reason ...

**Comment by jolivier23** [comment]
changed the title to csum since jira query will like that

**Review by jolivier23 — APPROVED** [review]

File: src/object/cli_obj.c [hunk 4]
```diff
@@ -4760,9 +4763,27 @@ obj_comp_cb(tse_task_t *task, void *data)
 						obj_auxi->tx_uncertain = 1;
 					else
 						obj_auxi->nvme_io_err = 1;
-				} else if (task->dt_result != -DER_NVME_IO) {
-					/* Don't retry update for CSUM & UNCERTAIN errors */
-					obj_auxi->io_retry = 0;
+				} else {
+					if (task->dt_result == -DER_CSUM) {
```
[the remark this answers is not in the document]
Comment (cdavis28) [reply src/object/cli_obj.c]: I see the "if" was resolved. What about the other comment?
```as for other operation, the "rw_arg = &obj_auxi->rw_args" pointer possibly invalid.```
with the extra check, does that protect invalid pointer access or is there another issue?

[the remark this answers is not in the document]
Comment (jolivier23) [reply src/object/cli_obj.c]: I think @liuxuezhao is saying that without limiting this to DAOS_OBJ_RPC_UPDATE, the field would be invald for other operations.  But with the if check, it prevents this.  Happy to let him clarify if I'm wrong.

**Review by cdavis28 — APPROVED** [review]

**Review by liuxuezhao — APPROVED** [review]

File: src/object/cli_obj.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (liuxuezhao) [reply src/object/cli_obj.c]: yes, the new version resolved the issue, thanks.

**Pull request merged**